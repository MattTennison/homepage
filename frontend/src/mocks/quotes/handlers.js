import { rest } from 'msw';
import response from './fixtures/response.json';

export const successfulHandler = rest.get('https://type.fit/api/quotes', (req, res, ctx) => {
    return res(
        ctx.status(200),
        ctx.json(response)
    )
});

export const errorHandler = rest.get('https://type.fit/api/quotes', (req, res, ctx) => {
    return res(
        ctx.status(404)
    )
});