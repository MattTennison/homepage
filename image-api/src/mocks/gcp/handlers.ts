import {
  DefaultRequestBody,
  RequestParams,
  ResponseResolver,
  rest,
  RestContext,
  RestRequest,
} from "msw";
import { clone } from "../../utils/clone";
import successfulSafeSearchFixture from "./fixtures/safesearch.json";

const makeSafeSearchResponse = (
  responseHandler: ResponseResolver<
    RestRequest<DefaultRequestBody, RequestParams>,
    RestContext,
    any
  >
) => {
  return rest.post(
    "https://vision.googleapis.com/v1/images:annotate",
    responseHandler
  );
};

export const successfulSafeSearchResponse = makeSafeSearchResponse(
  (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(successfulSafeSearchFixture));
  }
);

export const mangledSafeSearchResponse = makeSafeSearchResponse(
  (req, res, ctx) => {
    return res(ctx.status(200), ctx.json({ manged: "data" }));
  }
);

export const emptyResponsesSafeSearch = makeSafeSearchResponse(
  (req, res, ctx) => {
    const emptyResponses = {
      ...clone(successfulSafeSearchFixture),
      responses: [],
    };

    return res(ctx.status(200), ctx.json(emptyResponses));
  }
);

export const errorCodeSafeSearch = (errorCode: number) =>
  makeSafeSearchResponse((req, res, ctx) => {
    return res(ctx.status(errorCode));
  });
