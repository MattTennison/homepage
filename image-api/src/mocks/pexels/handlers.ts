import { FileHandle, readFile } from "fs/promises";
import {
  DefaultRequestBody,
  RequestParams,
  ResponseResolver,
  rest,
  RestContext,
  RestRequest,
} from "msw";
import { resolve } from "path";
import { URL } from "url";
import { dogPhoto } from "./fixtures/image-assets/helpers";
import successfulSearchResponse from "./fixtures/search.json";

const setupSearchResponse = (
  responseHandler: ResponseResolver<
    RestRequest<DefaultRequestBody, RequestParams>,
    RestContext,
    any
  >
) => {
  return rest.get("https://api.pexels.com/v1/search", responseHandler);
};

export const successfulSearch = setupSearchResponse((req, res, ctx) => {
  return res(ctx.status(200), ctx.json(successfulSearchResponse));
});

export const mangledSearch = setupSearchResponse((req, res, ctx) => {
  return res(ctx.status(200), ctx.json({ manged: "data" }));
});

export const errorCodeSearch = (errorCode: number) =>
  setupSearchResponse((req, res, ctx) => {
    return res(ctx.status(errorCode));
  });

const setupImageResponse = (
  responseHandler: ResponseResolver<
    RestRequest<DefaultRequestBody, RequestParams>,
    RestContext,
    any
  >
) => {
  return rest.get(
    "https://images.pexels.com/photos/:photoId/*",
    responseHandler
  );
};

export const assetResponse = (
  absolutePath: string | Buffer | URL | FileHandle
) =>
  setupImageResponse(async (req, res, ctx) => {
    const asset = await readFile(absolutePath);

    return res(
      ctx.status(200),
      ctx.set("Content-Length", asset.byteLength.toString()),
      ctx.set("Content-Type", "image/jpeg"),
      ctx.body(asset)
    );
  });

export const successfulAssetResponse = setupImageResponse(
  async (req, res, ctx) => {
    return res(ctx.status(200), ctx.body(dogPhoto.buffer));
  }
);

export const errorCodeAsset = (errorCode: number) =>
  setupImageResponse(async (req, res, ctx) => {
    return res(ctx.status(errorCode));
  });
