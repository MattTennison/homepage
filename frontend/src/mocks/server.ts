import { setupServer } from "msw/node";
import { successfulHandler } from "./quotes/handlers";

export const server = setupServer(successfulHandler);
