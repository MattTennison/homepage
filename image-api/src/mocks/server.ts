import { setupServer } from "msw/node";
import { successfulSafeSearchResponse } from "./gcp/handlers";
import { successfulAssetResponse, successfulSearch } from "./pexels/handlers";

export const server = setupServer(
  successfulSearch,
  successfulSafeSearchResponse,
  successfulAssetResponse
);
