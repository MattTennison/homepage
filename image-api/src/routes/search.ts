import { Observable } from "rxjs";
import { analyse } from "../services/gcp-safesearch";
import { fetchImageInBase64, search as pexelsSearch } from "../services/pexels";

type SearchResponse = {
  type: "ASSET";
  payload: string;
};

const checkImageForExplictContent = async (absolutePexelsPath: string) => {
  const base64 = await fetchImageInBase64({
    path: absolutePexelsPath.replace("https://images.pexels.com", ""),
  });
  const { safeSearch } = await analyse({ imageInBase64: base64 });

  const categoriesToCheck = ["adult", "violence", "racy"];
  const isKnownSafe = categoriesToCheck.every(
    (category) =>
      safeSearch[category] === "VERY_UNLIKELY" ||
      safeSearch[category] === "UNLIKELY"
  );

  return {
    isKnownSafe,
    payload: absolutePexelsPath,
  };
};

export const search = ({ query }: { query: string }) => {
  return new Observable<SearchResponse>((subscriber) => {
    pexelsSearch({ query }).then((assets) => {
      return Promise.allSettled(
        assets.photos.map((photo) =>
          checkImageForExplictContent(photo.src)
            .then((result) => {
              if (result.isKnownSafe) {
                subscriber.next({ type: "ASSET", payload: result.payload });
              }
            })
            .catch((err) => {
              console.log(err);
              throw err;
            })
        )
      ).then(() => subscriber.complete());
    });
  });
};
