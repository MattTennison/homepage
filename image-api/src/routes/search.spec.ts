import { filter, toArray } from "rxjs/operators";
import { safeSearchResponse } from "../mocks/gcp/handlers";
import { dogPhoto } from "../mocks/pexels/fixtures/image-assets/helpers";
import { server } from "../mocks/server";
import { search } from "./search";

jest.mock("../config", () => ({
  config: {
    ...jest.requireActual("../config").config,
    gcp: {
      apiToken: "GCP-token",
    },
    pexels: {
      apiToken: "pexels-api-token",
      maxImageSizeInBytes: 10 * 1024,
    },
  },
}));

describe("Search Route", () => {
  describe("with successful API responses", () => {
    beforeAll(() => {
      server.listen({ onUnhandledRequest: "error" });
    });

    afterEach(() => {
      server.resetHandlers();
    });

    afterAll(() => {
      server.close();
    });

    it("returns image URLs", (done) => {
      search({ query: "oceans" })
        .pipe(
          filter((value) => value.type === "ASSET"),
          toArray()
        )
        .subscribe((result) => {
          const urls = result.map((response) => response.payload);
          expect(result).toHaveLength(15);
          expect(urls).toMatchSnapshot();
          done();
        });
    });

    describe("SafeSearch filtering", () => {
      const categories = ["adult", "violence", "racy"];
      const unacceptableValues = [
        "UNKNOWN",
        "POSSIBLE",
        "LIKELY",
        "VERY_LIKELY",
      ];

      describe.each(categories)("for %s images", (category) => {
        it.each<string | jest.DoneCallback>(unacceptableValues)(
          "removes images with likelihood %s",
          (likelihood: string, done: jest.DoneCallback) => {
            server.use(safeSearchResponse({ [category]: likelihood }));

            search({ query: "oceans" })
              .pipe(
                filter((value) => value.type === "ASSET"),
                toArray()
              )
              .subscribe((result) => {
                expect(result).toBeEmpty();
                done();
              });
          }
        );
      });
    });
  });
});
