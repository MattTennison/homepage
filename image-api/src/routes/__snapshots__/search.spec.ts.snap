// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Route with successful API responses returns image URLs 1`] = `
Array [
  "https://images.pexels.com/photos/189349/pexels-photo-189349.jpeg?auto=compress&cs=tinysrgb&h=650&w=940",
  "https://images.pexels.com/photos/1001682/pexels-photo-1001682.jpeg?auto=compress&cs=tinysrgb&h=650&w=940",
  "https://images.pexels.com/photos/1680140/pexels-photo-1680140.jpeg?auto=compress&cs=tinysrgb&h=650&w=940",
  "https://images.pexels.com/photos/533923/pexels-photo-533923.jpeg?auto=compress&cs=tinysrgb&h=650&w=940",
  "https://images.pexels.com/photos/847393/pexels-photo-847393.jpeg?auto=compress&cs=tinysrgb&h=650&w=940",
  "https://images.pexels.com/photos/932638/pexels-photo-932638.jpeg?auto=compress&cs=tinysrgb&h=650&w=940",
  "https://images.pexels.com/photos/1802268/pexels-photo-1802268.jpeg?auto=compress&cs=tinysrgb&h=650&w=940",
  "https://images.pexels.com/photos/1533720/pexels-photo-1533720.jpeg?auto=compress&cs=tinysrgb&h=650&w=940",
  "https://images.pexels.com/photos/561463/pexels-photo-561463.jpeg?auto=compress&cs=tinysrgb&h=650&w=940",
  "https://images.pexels.com/photos/37403/bora-bora-french-polynesia-sunset-ocean.jpg?auto=compress&cs=tinysrgb&h=650&w=940",
  "https://images.pexels.com/photos/756856/pexels-photo-756856.jpeg?auto=compress&cs=tinysrgb&h=650&w=940",
  "https://images.pexels.com/photos/994605/pexels-photo-994605.jpeg?auto=compress&cs=tinysrgb&h=650&w=940",
  "https://images.pexels.com/photos/2860705/pexels-photo-2860705.jpeg?auto=compress&cs=tinysrgb&h=650&w=940",
  "https://images.pexels.com/photos/1452701/pexels-photo-1452701.jpeg?auto=compress&cs=tinysrgb&h=650&w=940",
  "https://images.pexels.com/photos/1076758/pexels-photo-1076758.jpeg?auto=compress&cs=tinysrgb&h=650&w=940",
]
`;
